### ISSUE TITLE
Copy large files forces system reboot

### ISSUE THREAD

**Issue by starfile** [openzfs/zfs#15411]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| macOS Ventura and Monterey
Distribution Version	| 13.6 and 12.7
Kernel Version	| Darwin 22.6.0 and 21.6.0
Architecture	| x86-64
OpenZFS Version	| 2.1.6

### Describe the problem you're observing
Copy large files using `rsync(1)` causes persistent high CPU consumption to two processes: `kernel_task` and `zed`. The last one is the `ZFS Event Daemon`. The issue starts between 1 to 1.7 TB at the specific file `mdbx.dat`. It is the ethereum chaindata with ~1.7 TB size. The process `kernel_task` has the highest CPU consumption and the process `zed` has higher CPU consumption than normal. The memory consumption for both processes showed no unusual behaviour.

### Describe how to reproduce the problem
A source disk (HDD) is formatted with APFS using Disk Utility app. The capacity is 8 TB and the used data size is 1.93 TB. Two target disks (SSDs) are formatted with ZFS using one pool and `lz4` compression. The capacity is 3.96 TB together. The issue was reproduced on two different machines with 32 GB and 16 GB RAM. All disks were connected by USB cables.

ZFS setup:

```
$ sudo zpool create tank disk6
$ sudo zpool add tank disk7
$ sudo zfs set compress=lz4 tank
```

Copy files:

```$ sudo rsync -av --delete SOURCE_PATH TARGET_PATH```

After few hours, the issue starts, see description in the section _Describe the problem you're observing_. In addition, the LEDs of the SSDs stopped blinking and the Activity Monitor shows that the `rsync(1)` process has ```0%``` CPU consumption. After few hours later, the `rsync(1)` command stops normal and claims that the copy task is done. The high CPU consumption of the `kernel_task` and `zed` processes keeps alive. ***A reboot is necessary***.

Checksum test:

```$ sudo rsync -avnc --delete SOURCE_PATH TARGET_PATH```

After few hours again, this test revealed that the copy task was successful. In addition, there was no issue with `kernel_task` or `zed` processes during the checksum test.

In summary, the issue must be somethings between the processes `zed` and `kernel_task` if write operation is involved.

**Comment by lundman** [comment]
Move: https://github.com/openzfsonosx/openzfs/issues

Have you checked if you get any *.panic files in /Library/Logs/DiagnosticLogs/ ?

**Comment by starfile** [comment]
> Have you checked if you get any *.panic files in /Library/Logs/DiagnosticLogs/ ?

There is not *.panic file in this folder.

I unplugged the target disk but kept the source disks with its zpool online. After few hours, the `zed` process started again to run on high CPU consumption without any copy task. After some `fs_usage` analysis, I see that there are lots of activity going on:

```
$ sudo fs_usage -f filesys PID 407
07:15:12.456040    WrData[AT1]     D=0x04c2ae16  B=0x1000   /dev/disk1s1  /private/var/run/zed.state                                                                                                                                                       0.000154 W zed.2186
07:15:12.457269  write             F=2    B=0x47                                                                                                                                                                                                           0.000028   zed.452117
07:15:12.458691  fcntl             F=4   <FULLFSYNC>                                                                                                                                                                                                       0.002825   zed.2186
07:15:12.462026  ioctl             F=5   <CMD=0xc0205a81>                                                                                                                                                                                                  0.000055   zed.2186
07:15:12.462129  write             F=2    B=0x10                                                                                                                                                                                                           0.000076   zed.2186
07:15:12.462161  open                   [  2] (RW__________)  /sys>>>>>>>>>>>>>>>>>>>>                                                                                                                                                                     0.000007   zed.2186
07:15:12.475683  write             F=2    B=0x30                                                                                                                                                                                                           0.000107   zed.2186
07:15:12.475809  write             F=2    B=0x46                                                                                                                                                                                                           0.000041   zed.452117
07:15:12.476672  write             F=2    B=0x31                                                                                                                                                                                                           0.000148   zed.2186
07:15:12.476729  lseek             F=4    O=0x00000000         <SEEK_SET>                                                                                                                                                                                  0.000003   zed.2186
07:15:12.476754  writev            F=4    B=0x18                                                                                                                                                                                                           0.000025   zed.2186
...
```
The output from `WrData[AT1]` to `F=4    B=0x18` repeats itself. The `/private/var/run/zed.state` output triggered me to look at the pool status:

```$ zpool status -x
  pool: tank
 state: ONLINE
status: One or more devices has experienced an unrecoverable error.  An
	attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
	using 'zpool clear' or replace the device with 'zpool replace'.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-9P
config:

	NAME                                          STATE     READ WRITE CKSUM
	tank                                           ONLINE       0     0     0
	  media-........-....-....-....-............   ONLINE       0  159M     0
	  media-........-....-....-....-............   ONLINE       0 79.6M     0
```
It seems that both source disks have errors. The error correction process seems to be very slow on those SSD disks, around 0.1 MB/s. I will recreate the zpool and try the whole copy process again.

**Comment by starfile** [comment]
The new recreated zpool did not solve the issue. I started the copy task again and `rsync(1)` copied successful. But the CPU consumption of the `kernel_task` and `zed` processes is again very high. Typing `zpool status -x` revealed that the source disks have again errors although they were made from scratch. I do not know what causes these errors. Both SSDs are brandnew and exactly the same model number: AGAMMIXS70B-2T-CS.

```
$ zpool status -x
  pool: tank
 state: ONLINE
status: One or more devices has experienced an unrecoverable error.  An
	attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
	using 'zpool clear' or replace the device with 'zpool replace'.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-9P
config:

	NAME                                          STATE     READ WRITE CKSUM
	tank                                           ONLINE       0     0     0
	  media-........-....-....-....-............   ONLINE       0 74.9M     0
	  media-........-....-....-....-............   ONLINE       0  150M     0

errors: No known data errors
```

**Comment by lundman** [comment]
That you are actually getting write errors is interesting, and makes me more concerned about heat than everything else. If you get lots of errors from the hardware. `zed` will be busy processing them which explains why it takes cpu.

What happens if you format it hfs and write a bunch of data to it?

**Comment by starfile** [comment]
You might be right about the heating cause. I used RAID Assistant from Disk Utility to create a striped RAID 0 on both SSDs using APFS rather than HFS. After few hours of write activity using erigon client, the client went stuck at high CPU consumption. The `fs_usage` utility revealed that the write activity has stopped and the content of the `/Volumes/` path revealed that somehow the original volume called `xpg` was remounted to `xpg 1`. Writing on `xpg 1` was possible but the erigon client denied to use the new mount point, claimed that the database file on the disk is corrupt. After reboot of macOS, the erigon client worked normally and continued its write activity. At that point, we know that something causes an auto remount and this could be the reason why `zed` does not work properly when lots of data get written to the disks.

Finally, I replaced the docking station (product number DA-71545) with two G-M2DK-AC-10G, each with integrated active cooling, and recreated the zpool on both SSDs to reproduce the issue. This time, the copy task was successful. Thus, I would agree with your assumption that the issue is caused by the heat development rather than ZFS. Interestingly, the initial hardware setup with only one SSD worked smooth at the beginning, 24x7 for two weeks. Once I plugged a second SSD into it, the issue began. Paradoxically, the manufacturer claims that its docking station is able to clone two SSDs, each up to 4 TB per slot.

I would say that this particular issue is solved. Many thanks for the advices!
